### PR TITLE
[MM-12361] Create CLI-command "config set" (#9535)

### PR DESCRIPTION
**Pull request by mukulrawat1986**
* Create the config set command in the variable

* Register the command and declare the command in the run function

* Finish implementation of 'config set' cli command

* Write tests for config set command

### CODE REVIEW

**Comment by mukulrawat1986** [comment]
@jespino please review when you have the time.

**Comment by mukulrawat1986** [comment]
Rebasing done

**Comment by grundleborg** [comment]
Will proceed with this one once https://github.com/mattermost/mattermost-server/pull/9564 is merged.

**Comment by grundleborg** [comment]
@mukulrawat1986 OK so the other two `config` CLI command PRs have both been merged, so we're ready to move onto this one. When you have a moment, please could you rebase to remove the conflicts.

**Comment by mukulrawat1986** [comment]
@grundleborg  Rebasing done

**Comment by grundleborg** [comment]
@mukulrawat1986 you've got a test failing on Jenkins, which looks related to this PR I think. The output is:

```
--- FAIL: TestSetConfig (0.53s)
    config_test.go:429: 
        	Error Trace:	config_test.go:429
        	Error:      	Received unexpected error:
        	            	exit status 1
        	Test:       	TestSetConfig
```

**Comment by mukulrawat1986** [comment]
@grundleborg  I have corrected the tests.

**Comment by mukulrawat1986** [comment]
@grundleborg can you see if this can be merged now?

File: cmd/mattermost/commands/config.go [hunk 1]
```diff
@@ -58,6 +59,15 @@ var ConfigShowCmd = &cobra.Command{
 	RunE:    configShowCmdF,
 }
 
+var ConfigSetCmd = &cobra.Command{
+	Use:     "set",
+	Short:   "Allow setting the value of a config setting",
```
Comment (hmhealey) [inline cmd/mattermost/commands/config.go]:
```
Short: "Set config setting",
Long: "Sets the value of a config setting by its name in dot notation. Accepts multiple values for array settings.",
```

File: cmd/mattermost/commands/config.go [hunk 2]
```diff
@@ -224,6 +235,145 @@ func printMap(value reflect.Value, tabVal int) string {
...
+
+		*update = *newConfig
+
+	}
+}
+
+func updateMap(configMap map[string]interface{}, configSettings []string, newVal []string) error {
+	res, ok := configMap[configSettings[0]]
+	if !ok {
+		return fmt.Errorf("config setting is not in the file")
```
Comment (hmhealey) [inline cmd/mattermost/commands/config.go]:
```suggestion
		return fmt.Errorf("Unable to find a setting with that name")
```

File: cmd/mattermost/commands/config.go [hunk 3]
```diff
@@ -224,6 +235,145 @@ func printMap(value reflect.Value, tabVal int) string {
...
+	}
+
+	value := reflect.ValueOf(res)
+
+	switch value.Kind() {
+
+	case reflect.Map:
+		// we can only change the value of a particular setting, not the whole map, return error
+		if len(configSettings) == 1 {
+			return fmt.Errorf("can't change the setting of the whole map")
```
Comment (hmhealey) [inline cmd/mattermost/commands/config.go]:
```suggestion
			return fmt.Errorf("Unable to set multiple settings at once")
```

File: cmd/mattermost/commands/config.go [hunk 4]
```diff
@@ -224,6 +235,145 @@ func printMap(value reflect.Value, tabVal int) string {
...
+	case reflect.Int:
+		if len(configSettings) == 1 {
+			val, err := strconv.Atoi(newVal[0])
+			if err != nil {
+				return err
+			}
+			configMap[configSettings[0]] = val
+			return nil
+		}
+		return fmt.Errorf("Config setting not in file")
```
Comment (hmhealey) [inline cmd/mattermost/commands/config.go]: For all of these,
```suggestion
		return fmt.Errorf("Unable to find a setting with that name")
```

File: cmd/mattermost/commands/config.go [hunk 5]
```diff
@@ -224,6 +235,145 @@ func printMap(value reflect.Value, tabVal int) string {
...
+
+func updateMap(configMap map[string]interface{}, configSettings []string, newVal []string) error {
+	res, ok := configMap[configSettings[0]]
+	if !ok {
+		return fmt.Errorf("config setting is not in the file")
+	}
+
+	value := reflect.ValueOf(res)
+
+	switch value.Kind() {
```
Comment (hmhealey) [inline cmd/mattermost/commands/config.go]: Does this work for the config settings that are pointers? I'm not sure how the `value.Kind()` works for those

**Comment by mukulrawat1986** [comment]
@hmhealey Made the changes that you have requested. As for the pointer issue I have replied on your comment.

**Comment by mukulrawat1986** [comment]
@grundleborg Sure, I will add the unit tests for `updateMap()` by tomorrow

**Comment by mukulrawat1986** [comment]
@grundleborg I have written the unit test for UpdateMap, please review.

**Pull request merged**